codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/Paginators.java

```diff
@@ -0,0 +1,254 @@
+
+                    List<MemberShape> outputMemberPath = paginationInfo.getOutputTokenPath();
+                    for (int i = 0; i < outputMemberPath.size(); i++) {
+                        MemberShape memberShape = outputMemberPath.get(i);
+                        outputPath.append(".");
+                        outputPath.append(symbolProvider.toMemberName(memberShape));
+
+                        // Don't need to check the last member here as we won't dereference it
+                        if (i != outputMemberPath.size() - 1) {
+                            if (i != 0) {
+                                nilGuard.append(" && ");
+                            }
+                            nilGuard.append(outputPath);
+                            nilGuard.append(" != nil");
+                        }
+                    }
+
+                    Runnable setToken = () -> {
+                        writer.write("p.nextToken = $L", outputPath);
+                    };
+                    if (nilGuard.length() > 0) {
+                        writer.openBlock("if $L {", "}", nilGuard.toString(), setToken::run);
+                    } else {
+                        setToken.run();
+                    }
+
+                    writer.openBlock("if p.options.StopOnDuplicateToken && "
+                            + "params.$L != nil && p.nextToken != nil && "
+                            + "*params.$L == *p.nextToken {", "", inputMember, inputMember, () -> {
+                        writer.write("p.done = true");
+                        writer.openBlock("} else {", "}", () -> writer.write("p.done = false"));
+                    });
+
+                    writer.write("");
+                    writer.write("return result, nil");
+                });
+    }
+
+    private void writePaginatorOptions(
+            Model model,
+            SymbolProvider symbolProvider,
+            GoWriter writer,
+            Symbol operationSymbol,
+            Optional<MemberShape> pageSizeMember,
+            Symbol optionsSymbol
+    ) {
+        writer.writeDocs(String.format("%s is the paginator options for %s", optionsSymbol.getName(),
+                operationSymbol.getName()));
+        writer.openBlock("type $T struct {", "}", optionsSymbol, () -> {
+            pageSizeMember.ifPresent(memberShape -> {
+                MemberShape limitShape = pageSizeMember.get();
+                limitShape.getMemberTrait(model, DocumentationTrait.class).ifPresent(documentationTrait -> {
+                    writer.writeDocs(documentationTrait.getValue());
+                });
+                writer.write("Limit $P", symbolProvider.toSymbol(limitShape));
+                writer.write("");
+            });
+            writer.writeDocs("Set to true if pagination should stop if the service returns "
+                    + "the same taken passed to it");
+            writer.write("StopOnDuplicateToken bool");
```

Comment:
Minor nit, with doc.  Maybe something to clarify that the same token returned for subiquent requests. Basically to clarify that the sdk won't keep a history of tokens, and only compares against the previous and most recent.